### PR TITLE
ISPN-14088 Endpoint to expose cache configuration default values

### PR DESCRIPTION
**Pull request by dpanshug**
https://issues.redhat.com/browse/ISPN-14088

### CODE REVIEW

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 1]
```diff
@@ -466,6 +473,30 @@ private CompletionStage<RestResponse> dataSourceTest(RestRequest restRequest) {
...
+      return asJsonResponseFuture(Json.make(attributes), isPretty(request));
+   }
+
+   private static void allAttributes(ConfigurationElement<?> element, Map<String, String> attributes) {
+      for (Attribute<?> attribute : element.attributes().attributes()) {
+            AttributeDefinition<?> definition = attribute.getAttributeDefinition();
+            String value;
+            try{
+               value = attribute.getInitialValue().toString();
+            } catch (Exception e){
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:489]: We should just be able to check if the initial value is null or not and not catch all Exception.

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 2]
```diff
@@ -466,6 +473,30 @@ private CompletionStage<RestResponse> dataSourceTest(RestRequest restRequest) {
...
+      Map<String, String> attributes = new LinkedHashMap<>();
+      allAttributes(configuration, attributes);
+      return asJsonResponseFuture(Json.make(attributes), isPretty(request));
+   }
+
+   private static void allAttributes(ConfigurationElement<?> element, Map<String, String> attributes) {
+      for (Attribute<?> attribute : element.attributes().attributes()) {
+         AttributeDefinition<?> definition = attribute.getAttributeDefinition();
+         String value;
+         if(attribute.getInitialValue() == null){
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:487]: I would recommend importing the IntelliJ style code from https://github.com/infinispan/infinispan/tree/main/ide-settings so it will automatically fix formatting things such as this line.

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 3]
```diff
@@ -466,6 +473,30 @@ private CompletionStage<RestResponse> dataSourceTest(RestRequest restRequest) {
...
+            value = null;
+         } else {
+            value = attribute.getInitialValue().toString();
+         }
+         attributes.put(definition.name(), value);
+      }
+      for (ConfigurationElement<?> child : element.children()) {
+         allAttributes(child, attributes);
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:495]: I just noticed this is recursively calling itself, however we don't have a prefix or anything. Are we sure no property is being overridden by a sub element? For example if we have a "timeout" setting that is across multiple attribute definitions only the last will show and not know which it applies to.

**Review by wburns — COMMENTED** [review]
More for clarification and a tip to make sure you are using the code style imports

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java (hunk 3 above)
Comment (dpanshug) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java]: I had forgotten about this case, but I have now updated the PR to reflect the necessary changes.

Now the API will output a JSON like: 
```
{
  "clustering": {
    "mode": "LOCAL",
    "remote-timeout": "15000",
    ...
  },
  "hash": {
    "consistent-hash-factory": null,
    "owners": "2",
   ...
  },
  ...
}
```

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 4]
```diff
@@ -466,6 +474,33 @@ private CompletionStage<RestResponse> dataSourceTest(RestRequest restRequest) {
...
+         AttributeDefinition<?> definition = attribute.getAttributeDefinition();
+         String value;
+         if (attribute.getInitialValue() == null) {
+            value = null;
+         } else {
+            value = attribute.getInitialValue().toString();
+         }
+         attributeMap.put(definition.name(), value);
+      }
+      attributes.put(prefix, attributeMap);
```
Comment (jabolina) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: Should we return something representing how it is nested in the configuration? IIUC, right now we have everything directly under the JSON root.

Comment (dpanshug) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: @jabolina can you elaborate more, what and where exactly you want to represent?

Comment (jabolina) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: Maybe what I am saying is unnecessary. For example, key and value encodings are nested under the encoding configuration:

```java
new ConfigurationBuilder()
    .encoding()
        .key().mediaType(...)
    .encoding()
        .value().mediaType(...)
    .build();
```

But on the JSON, we have that:

```json
{
   "encoding":{
      "media-type": null
   },
   "key":{
      "media-type": null
   },
   "value":{
      "media-type": null
   }
}
```

Everything is under the root, it doesn't nest in the same way when configuring.

Comment (wburns) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: The question for me is can I copy/paste these defaults and put it in to create a cache. This would ensure the layout is the same and that we know a config maps to the one it says it is.

Comment (dpanshug) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: > But on the JSON, we have that:

@jabolina This nesting pattern for encoding we get is from `ConfigurationBuilder().build();`

Comment (dpanshug) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: > The question for me is can I copy/paste these defaults and put it in to create a cache. This would ensure the layout is the same and that we know a config maps to the one it says it is.

I don't think this JSON of default values can be considered for creating a cache. For a cache configuration we need to have cache mode at the start of configuration, which can't be possible in this case.

Comment (wburns) [reply server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:497]: The defaults are literally what is created when the cache is just defined and not overridden though, so it better be able to define a cache with.

**Comment by dpanshug** [comment]
rebased

**Comment by karesti** [comment]
@jabolina @wburns is there anything else needed here ? 🙏🏻

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 5]
```diff
@@ -459,6 +472,33 @@ private CompletionStage<RestResponse> dataSourceTest(RestRequest restRequest) {
...
    }
 
+   private CompletionStage<RestResponse> getCacheConfigDefaultAttributes(RestRequest request) {
+      Configuration configuration = new ConfigurationBuilder().build();
+      resetAll(configuration);
+      StringBuilderWriter sw = new StringBuilderWriter();
+      ConfigurationBuilderHolder holder = invocationHelper.getParserRegistry().parse(new String(bytes, UTF_8), APPLICATION_JSON);
+      ConfigurationBuilder cfgBuilder = holder.getCurrentConfigurationBuilder() != null ? holder.getCurrentConfigurationBuilder() : new ConfigurationBuilder();
+      GlobalConfiguration globalConfiguration = SecurityActions.getCacheManagerConfiguration(invocationHelper.getRestCacheManager().getInstance());
+      System.out.println("globalConfiguration = " + globalConfiguration);
```
Comment (jabolina) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:482]: These are unused and can be removed. As well as the println below.

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 6]
```diff
@@ -31,13 +32,18 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
+//import java.util.Map;
```
Comment (jabolina) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:35]: Can remove all these commented imports.

File: server/tests/src/test/java/org/infinispan/server/functional/rest/RestServerResource.java [hunk 7]
```diff
@@ -93,4 +93,15 @@ public void testCacheManagerNames() {
       Set<String> cmNames = cacheManagers.asJsonList().stream().map(Json::asString).collect(Collectors.toSet());
       assertEquals(cmNames, Sets.newHashSet("default"));
    }
+
+   @Test
+   public void testCacheDefaults() {
```
Comment (jabolina) [inline server/tests/src/test/java/org/infinispan/server/functional/rest/RestServerResource.java:98]: I believe it is worth validating more fields here. And another good test to have would be to retrieve the defaults and create a new cache with the return.

**Review by jabolina — CHANGES_REQUESTED** [review]

**Comment by jabolina** [comment]
Replaying CI.

Just to clarify, this returns the default values, does the return needs to be a valid configuration?

**Comment by karesti** [comment]
default values @jabolina

**Comment by jabolina** [comment]
It looks like the endpoint was removed. Since it is only necessary to return the default values without being a valid configuration, you can take the approach you were on initially, @dpanshug.

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java [hunk 8]
```diff
@@ -180,6 +180,13 @@ public AttributeSet protect() {
       return protectedSet;
    }
 
+   public void asModified() {
```
Comment (jabolina) [inline commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java:183]: You can revert these changes as they are not necessary anymore.

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java [hunk 9]
```diff
@@ -61,6 +61,10 @@ public Attribute<T> protect() {
       return this;
    }
 
+   void asModified() {
```
Comment (jabolina) [inline commons/all/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java:64]: Here, too.

**Review by jabolina — COMMENTED** [review]
Looking good, just need to revert some changes and we're good.

**Comment by jabolina** [comment]
I think its looking good :smile: :smile:

```json
{
  "local-cache": {
    "configuration": null,
    "simple-cache": "false",
    "clustering": {
      "mode": "LOCAL",
      "remote-timeout": "15000",
      "hash": {
        "consistent-hash-factory": null,
        "owners": "2",
        "segments": "256",
        "capacity-factor": "1.0",
        "hash-function": "MurmurHash3",
        "key-partitioner": "HashFunctionPartitioner{hashFunction=MurmurHash3, ns=256}"
      },
      "l1": {
        "enabled": "false",
        "invalidation-threshold": "0",
        "l1-lifespan": "600000",
        "l1-cleanup-interval": "60000"
      },
      "state-transfer": {
        "enabled": "true",
        "timeout": "240000",
        "chunk-size": "512",
        "await-initial-transfer": "true"
      },
      "partition-handling": {
        "when-split": "ALLOW_READ_WRITES",
        "merge-policy": "NONE"
      }
    },
    "expiration": {
      "lifespan": "-1",
      "max-idle": "-1",
      "reaperEnabled": "true",
      "interval": "60000",
      "touch": "SYNC"
    },
    "encoding": {
      "media-type": null,
      "key": {
        "media-type": null
      },
      "value": {
        "media-type": null
      }
    },
    "query": {
      "properties": "{}",
      "default-max-results": "100"
    },
    "indexing": {
      "properties": "{}",
      "key-transformers": "{}",
      "indexed-entities": "[]",
      "enabled": "false",
      "storage": "filesystem",
      "startup-mode": "NONE",
      "path": null,
      "indexing-mode": "AUTO"
    },
    "jmx-statistics": {
      "statistics": "false",
      "statistics-available": "true"
    },
    "locking": {
      "concurrency-level": "32",
      "isolation": "REPEATABLE_READ",
      "acquire-timeout": "10000",
      "striping": "false"
    },
    "transaction": {
      "auto-commit": "true",
      "stop-timeout": "30000",
      "locking": "OPTIMISTIC",
      "transaction-manager-lookup": "org.infinispan.transaction.lookup.GenericTransactionManagerLookup@44724d1d",
      "transaction-synchronization-registry-lookup": null,
      "mode": "NON_TRANSACTIONAL",
      "synchronization": "false",
      "single-phase-auto-commit": "false",
      "reaper-interval": "30000",
      "complete-timeout": "60000",
      "notifications": "true",
      "recovery": {
        "enabled": "false",
        "recovery-cache": "__recoveryInfoCacheName__"
      }
    },
    "unsafe": {
      "unreliable-return-values": "false"
    },
    "sites": {
      "merge-policy": "DEFAULT",
      "max-cleanup-delay": "30000",
      "tombstone-map-size": "512000",
      "backups": {},
      "backup-for": {
        "remote-cache": null,
        "remote-site": null
      }
    },
    "memory": {
      "storage": "HEAP",
      "max-size": null,
      "max-count": "-1",
      "when-full": "NONE"
    }
  }
}
```

**Comment by jabolina** [comment]
Integrated into main, thanks @dpanshug!
